L58-python-async-api.md

```diff
@@ -0,0 +1,1708 @@
+new sets of async APIs.
+
+### Flow Control Enforcement
+
+To propagate HTTP/2 flow control push back, the new async API needs to aware of
+the flow control mechanism. Most complex logic is handled in C-Core, except
+there is a single rule the wrapper layer needs to follow: there can only be one
+outstanding read/write operation on each call.
+
+It means if the application fires two write in parallel, one of them have to
+wait until the other one finishes. Also, that rule doesn't prohibit reading and
+writing at the same time.
+
+So, even if all the read/write is asynchronous in `asyncio`, we will have to
+either enforce the rule ourselves by adding locks in our implementation. Or we
+can pass down the synchronization responsibility to our users.
+
+### Concrete Class Instead of "Interfaces"
+
+Interface is a design pattern that defines the contract of an entity that allows
+different implementation to work seamlessly in a system. In the past, gRPC
+Python has been using metaclass based Python interface pattern. It works just
+like interface in Golang and Java, except the error is generated in runtime
+instead of compile time.
+
+If the gRPC Python has multiple implementation for a single interface, the use
+of the design pattern provides productivity in unifying their behavior. However,
+almost non interfaces has second implementation, even if they do, they are
+depends directly on our concrete implementation, which should better using
+inheritance or composition than interfaces.
+
+Also, in the past, dependents of gRPC Python have observed several failure
+caused by the interface. The interface constraints our ability to add
+experimental API. Once we change even slightly with interfaces, the downstream
+implementations are likely to break.
+
+Since this is a new opportunity for us to re-design, we need to think cautiously
+about how do we empower our users to extend our classes. For majority of cases,
+we are providing the only implementation for the interface. We should
+convert them into concrete classes.
+
+On the other hand, there are actually one valid use case that we should keep
+abstract class -- interceptors. To be more specific, the following interfaces
+won't be replaced by concrete classes:
+
+* `grpc.ServerInterceptor`
+* `grpc.UnaryUnaryClientInterceptor`
+* `grpc.UnaryStreamClientInterceptor`
+* `grpc.StreamUnaryClientInterceptor`
+* `grpc.StreamStreamClientInterceptor`
+
+## Future Features
+
+### Implicit Context Propagation By `contextvars`
+
+`contextvars` is a Python 3.7 feature that allows applications to set coroutine
+local variables (just like thread local variables). In the past, gRPC Python
+uses thread local to achieve two goals:
+
+* Support distributed tracing library that work across languages;
+* Support deadline propagation along the chain of RPCs.
+
+For distributed tracing library (like OpenCensus), the tracing information is
+preserved as one of the metadata, and it is expected to be implicitly promoted
```

Comment:
Tracing should *not* be propagated by default because gRPC shouldn't have an opinion on what tracing instrumentation looks like. This needs to be pluggable. Instead, there needs to be an API by which individual tracing library authors may write their own middleware interceptors that determine how to interpret metadata into tracing data and how to convert tracing data to metadata. I will follow up with an example in grpc-api-examples on this.